opensearch/src/main/java/org/opensearch/sql/opensearch/data/type/OpenSearchDateType.java

```diff
@@ -0,0 +1,144 @@
+import static org.opensearch.sql.data.type.ExprCoreType.UNKNOWN;
+
+import com.google.common.collect.ImmutableMap;
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeFormatterBuilder;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+import java.util.Map;
+import java.util.stream.Collectors;
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.Setter;
+import org.opensearch.common.time.DateFormatter;
+import org.opensearch.sql.data.type.ExprType;
+
+/**
+ * Of type join with relations. See
+ * <a href="https://opensearch.org/docs/latest/opensearch/supported-field-types/join/">doc</a>
+ */
+@EqualsAndHashCode(callSuper = true)
+public class OpenSearchDateType extends OpenSearchDataType {
+
+  private static final OpenSearchDateType instance = new OpenSearchDateType();
+
+
+  // a read-only collection of relations
+  @Getter
+  @Setter
+  @EqualsAndHashCode.Exclude
+  DateTimeFormatter formatter;
+
+  @Getter
+  @Setter
+  @EqualsAndHashCode.Exclude
+  String formatString;
+
+  private OpenSearchDateType() {
+    super(MappingType.Timestamp);
+    this.formatter = null;
+    this.formatString = "";
+  }
+
+  private OpenSearchDateType(DateTimeFormatter formatterArg, String formatStringArg, MappingType mappingType) {
+    super(mappingType);
+    this.formatter = formatterArg;
+    this.formatString = formatStringArg;
+  }
+
+  public static List<String> getFormatList(String formats) {
+    if (formats == null || formats.isEmpty()) {
+      return List.of();
+    }
+    return Arrays.stream(formats.split("\\|\\|")).map(String::trim).collect(Collectors.toList());
+  }
+
+  public List<DateFormatter> getNamedFormatters(String formats) {
+    return getFormatList(formats).stream().filter(f -> {
+          try {
+            DateTimeFormatter.ofPattern(f);
+            return false;
+          } catch (Exception e) {
+            return true;
+          }
+        })
+        .map(DateFormatter::forPattern).collect(Collectors.toList());
+  }
+
+  public List<DateTimeFormatter> getRegularFormatters() {
+    return getFormatList(formatString).stream().map(f -> {
+          try {
+            return DateTimeFormatter.ofPattern(f);
+          } catch (Exception e) {
+            return null;
+          }
+        })
+        .filter(Objects::nonNull).collect(Collectors.toList());
+  }
+
+  /**
+   * Create a Date type which has a LinkedHashMap defining all formats
+   * @return A new type object.
+   */
+  public static OpenSearchDateType create(String format, MappingType mappingType) {
+    //TODO: Filter out named formatters vs user defined here
+    // Initialize the format based on the given string
+    DateTimeFormatterBuilder builder = new DateTimeFormatterBuilder();
+
+    if (format.contains("||")) {
+      for (String token: format.split("\\|\\|")) {
+        //try to append a pattern
+        try {
+          builder.appendPattern(token);
```

Comment:
Rework: format could be a mix of custom and pre-defined formats:
```
hour_minute_second||t_time
```

Reply:
This has been removed